Remount transactional data source

itsthejb opened the pull request:
Resolves https://github.com/facebook/componentkit/issues/342

As noted in the issue, I encountered the same remounting issue with the new data source as encountered in the original (https://github.com/facebook/componentkit/pull/256). The first part of the problem was to use the same technique of grabbing any existing cell for the index path and reusing it. Then an additional problem *seemed* to be that changes were applied to the collection view before updating the local state in `CKCollectionViewTransactionalDataSource`. Simply moving the `applyChangesToCollectionView()` call until after that fixed this.

So... this appears at quick inspection to fix the issue although I'm certainly not 100% confident about the ramifications. I also thought about migrating the example app to the new data source to illustrate the issue, but as far as I can remember it only uses `-updateState`, and not changeset modifications.

Anyway, looking forward to feedback :+1:

adamjernst commented:
@majak @benlodotcom can you guys review/merge this?

itsthejb commented:
I'm also looking into writing some tests - at least to demonstrate the issue. That might be a reasonably large job, though.

itsthejb commented:
We're also getting a lot of `attempt to create view animation for nil view` exceptions. Still haven't confirmed if it's not something that we're doing strange (although I don't *think* so). It's not related to this change, though - have tested with master.

itsthejb commented:
Updated with a test to demonstrate and issue and fix. Essentially want to test that the controller receives the `willRemount` and `didRemount` messages. Doing this is a subclass and posting notifications is a *little* funky, but it's the simplest strategy I could think of for the time being. Please let me know if there's some debugging stuff I could use that would make it a little easier to get to the component controller.

benlodotcom commented on ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm:
```diff
@@ -61,9 +61,14 @@ - (void)applyChangeset:(CKTransactionalComponentDataSourceChangeset *)changeset
+static void applyChangesToCollectionView(CKTransactionalComponentDataSourceAppliedChanges *changes, UICollectionView *collectionView, CKTransactionalComponentDataSourceState *currentState, CKComponentDataSourceAttachController *attachController)
 {
-  [collectionView reloadItemsAtIndexPaths:[changes.updatedIndexPaths allObjects]];
+  NSMutableArray *reloadIndexPaths = @[].mutableCopy;
```
This is a bit inefficient as it incurs a full copy of the array, can we do `[NSMutableArray array]` instead ?

benlodotcom commented on ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm:
```diff
@@ -61,9 +61,14 @@ - (void)applyChangeset:(CKTransactionalComponentDataSourceChangeset *)changeset
+  NSMutableArray *reloadIndexPaths = @[].mutableCopy;
+  [changes.updatedIndexPaths enumerateObjectsUsingBlock:^(NSIndexPath *indexPath, BOOL *stop) {
+    CKCollectionViewDataSourceCell *cell = (CKCollectionViewDataSourceCell*) [collectionView cellForItemAtIndexPath:indexPath];
+    cell ? _attachToCell(cell, indexPath, currentState, attachController) : [reloadIndexPaths addObject:indexPath];
```
If we don't have a cell it only means it's not visible or will be created when we apply changes to the CV, no need to reloadIndexPaths I think.

benlodotcom commented:
Thanks @itsthejb, I added some comments inline the reload is normally not necesssary.

itsthejb commented:
Looks good. I'll look into those changes when I have a moment.

itsthejb commented on ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm:
```diff
@@ -64,9 +64,13 @@ - (void)applyChangeset:(CKTransactionalComponentDataSourceChangeset *)changeset
+	if (CKCollectionViewDataSourceCell *cell = (CKCollectionViewDataSourceCell*) [collectionView cellForItemAtIndexPath:indexPath]) {
+		_attachToCell(cell, indexPath, currentState, attachController);
+	}
+  }];
```
Ok, provided I haven't misunderstood @benlodotcom's comment, there's actually no need for the mutable array at all, or doing a reload on any cells that aren't currently visible.

ocrickard replied:
@rnystrom at this point I think you're probably the best one to help review this. Any chance you mind taking a look?

itsthejb commented:
Ok, updated with recommended changes. Did this without Xcode available, but seems fine on Travis.

itsthejb commented:
Hi guys - seems this PR has been languishing for some time, even though we've been using it in production code with no issues for some time...

adamjernst commented:
Sorry, ack! I promise I'll take a look soon.

ocrickard commented:
@adamjernst ping on review here?

ocrickard commented on ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm:
```diff
@@ -134,11 +138,16 @@ - (void)updateConfiguration:(CKTransactionalComponentDataSourceConfiguration *)c
 
 static NSString *const kReuseIdentifier = @"com.component_kit.collection_view_data_source.cell";
 
+static void _attachToCell(CKCollectionViewDataSourceCell *cell, NSIndexPath *indexPath, CKTransactionalComponentDataSourceState *currentState, CKComponentDataSourceAttachController *attachController)
```
nit: we try not to prefix static functions with _

ocrickard commented on ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm:
```diff
@@ -65,9 +65,13 @@ - (void)applyChangeset:(CKTransactionalComponentDataSourceChangeset *)changeset
 {
-  [collectionView reloadItemsAtIndexPaths:[changes.updatedIndexPaths allObjects]];
+  [changes.updatedIndexPaths enumerateObjectsUsingBlock:^(NSIndexPath *indexPath, BOOL *stop) {
+	if (CKCollectionViewDataSourceCell *cell = (CKCollectionViewDataSourceCell*) [collectionView cellForItemAtIndexPath:indexPath]) {
```
nit: space before *: `CKCollectionViewDataSourceCell *`

ocrickard commented on ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm:
```diff
@@ -85,12 +89,12 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
                                                     inState:previousState];
     // Update current state
     _currentState = [_componentDataSource state];
+    applyChangesToCollectionView(changes, _collectionView, _currentState, _attachController);
```
You should add a comment here explaining the ordering.

ocrickard commented:
Alright, I'm just now reading this, and if I understand the problem:

1. We're calling `reload...` for the index paths that have updated.
2. This gives us a *new* cell, which causes us to receive didUnmount/didMount.

The change in the ordering of collection view mutation sequencing mentioned here:

> Then an additional problem seemed to be that changes were applied to the collection view before updating the local state in CKCollectionViewTransactionalDataSource. Simply moving the applyChangesToCollectionView() call until after that fixed this.

This makes sense, since in practice, the reload occurs when the batch is applied, after the block is invoked. Since now you're directly grabbing the current component and re-mounting it in the cell, you're gonna get the old state. The order changes do make sense to me, and to my knowledge should not affect anything.

I'm in favor of merging this PR, but would like @adamjernst to look as well since it touches such a sensitive code path.

itsthejb commented:
Yes, indeed! Well we've been using this in production for some time and no issues. It would certainly be nice to ditch our fork :) 

The unit test is a little funky, but does verify the issue... I await further feedback, anyway!

facebook-github-bot-5 commented:
@itsthejb I tried to find reviewers for this pull request and wanted to ping them to take another look. However, based on the blame information for the files in this pull request I couldn't find any reviewers. This sometimes happens when the files in the pull request are new or don't exist on master anymore. Is this pull request still relevant? If yes could you please rebase? In case you know who has context on this code feel free to mention them in a comment (one person is fine). Thanks for reading and hope you will continue contributing to the project.

itsthejb commented:
Rebased and updated with recommended changes. I added the comment as requested. Hope that's descriptive enough; this isn't very fresh in my mind anymore - all the more reason to have written a comment originally, I suppose...

itsthejb commented:
I believe tests are failing because `.travis.yml` doesn't run `carthage update`... But perhaps my mistake? It tests correctly locally with `carthage update; sh build.sh` (if a 9.3 iPhone 5 simulator is installed). I'm a little behind on general updates to the project structure / build workflow.

ocrickard commented:
Argh, yeah you may be right about the travis config. I don't understand it at all. Mind looking into it?

itsthejb commented:
Having a look at that here https://github.com/facebook/componentkit/pull/612

ocrickard commented:
Merged #612, mind rebasing so we can kick off the CI build again?

itsthejb commented:
Done 😉

ocrickard commented:
I'm in favor of this. I will merge it after we merge a few critical fixes so we can pick them in internally separately should this need to be reverted.

ocrickard commented on ComponentKitTests/TransactionalDataSource/CKTransactionalComponentDataSourceIntegrationTests.mm:
```diff
@@ -0,0 +1,123 @@
+}
+- (void)didRemount {
+  [super didRemount];
+  [_notificationCenter postNotificationName:NSStringFromSelector(_cmd) object:self];
```
Tests are a bit unorthodox here, but better than no tests I guess.

itsthejb replied:
Oh for sure! I'm open to ideas for a rewrite. The "problem" is that, since the controllers are entirely uncoupled, there's no more direct way to access them. This is kind of just the first thing that came to mind...

ocrickard commented:
Alright, we did all the high-pri merges we wanted, so now we will merge this and I will conduct some tests internally to ensure this doesn't break any of our advanced usages of components.

ocrickard commented:
This caused flakiness after merging. We are reverting. If you'd like us to consider re-merging please make sure you improve the tests.
